sale_order_type/models/sale.py

```diff
@@ -0,0 +1,115 @@
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+# Copyright 2020 Tecnativa - Pedro M. Baeza
+
+from odoo import api, fields, models
+
+
+class SaleOrder(models.Model):
+    _inherit = "sale.order"
+
+    type_id = fields.Many2one(
+        comodel_name="sale.order.type",
+        string="Type",
+        compute="_compute_sale_type_id",
+        store=True,
+        readonly=True,
+        states={
+            "draft": [("readonly", False)],
+            "sent": [("readonly", False)],
+        },
+        default=lambda so: so._default_type_id(),
+        ondelete="restrict",
+        copy=True,
+    )
+
+    @api.model
+    def _default_type_id(self):
+        return self.env["sale.order.type"].search([], limit=1)
+
+    @api.depends("partner_id", "company_id")
+    def _compute_sale_type_id(self):
+        for record in self:
+            if not record.partner_id:
+                record.type_id = self.env["sale.order.type"].search(
+                    [("company_id", "in", [self.env.company.id, False])], limit=1
+                )
+            else:
+                sale_type = (
+                    record.partner_id.with_company(record.company_id).sale_type
+                    or record.partner_id.commercial_partner_id.with_company(
+                        record.company_id
+                    ).sale_type
+                )
+                if sale_type:
+                    record.type_id = sale_type
+
+    @api.onchange("type_id")
+    def onchange_type_id(self):
+        # TODO: To be changed to computed stored readonly=False if possible in v14?
+        vals = {}
+        for order in self:
+            order_type = order.type_id
+            # Order values
+            vals = {}
+            if order_type.warehouse_id:
+                vals.update({"warehouse_id": order_type.warehouse_id})
+            if order_type.picking_policy:
+                vals.update({"picking_policy": order_type.picking_policy})
+            if order_type.payment_term_id:
+                vals.update({"payment_term_id": order_type.payment_term_id})
+            if order_type.pricelist_id:
+                vals.update({"pricelist_id": order_type.pricelist_id})
+            if order_type.incoterm_id:
+                vals.update({"incoterm": order_type.incoterm_id})
+            if vals:
+                order.update(vals)
+            # Order line values
+            line_vals = {}
+            line_vals.update({"route_id": order_type.route_id.id})
+            order.order_line.update(line_vals)
+
+    @api.model
+    def create(self, vals):
+        if vals.get("name", "/") == "/" and vals.get("type_id"):
+            sale_type = self.env["sale.order.type"].browse(vals["type_id"])
+            if sale_type.sequence_id:
+                vals["name"] = sale_type.sequence_id.next_by_id()
```

Comment:
I think this module is not working properly with the subsequences by date range.
I configure the subsequences for sale orders and, when I create once for the next year, it assings the 2021 sequence.

![sale_order_type_range_sequence](https://user-images.githubusercontent.com/29223264/124739060-ceb3d300-df19-11eb-8971-c13a8866b181.gif)

Reply:
ping @BT-cserra

Reply:
@ioans73 did you test with V13 or with this PR?

Reply:
@BT-cserra with this PR

Reply:
Many thanks @ioans73 you were 100% right! There was a missing parameter. [This commit](https://github.com/OCA/sale-workflow/pull/1646/commits/d53ccc6af7a5d70910c01c28bd49d843dc58d523) fixes it. I've requested a re-review.

Reply:
Wait @ioans73 I forgot to pass the tests. Ouch.

Reply:
I force pushed with https://github.com/OCA/sale-workflow/pull/1646/commits/d53ccc6af7a5d70910c01c28bd49d843dc58d523. Now the checks are in green.

Reply:
Perfect, thank you very much @BT-cserra